### PR TITLE
add mt-simulate-memory-idx-lock-contention tool + benchmark scripts

### PR DESCRIPTION
**Pull request by Dieterbe**
this tool allows to simulate a workload on the memory-idx, by preloading it with data, and then doing a benchmark stresstest in which we simultaneously issue queries, adds and updates, all at configurable rates. and measure the durations of each operation.

edit: now also includes a benchmark scripts to try different parameters and plot the results using gnuplot

### CODE REVIEW

**Comment by Dieterbe** [comment]
## writequeue effect on queries
writeq-false-query
writeq-true-query

## writequeue effect on metric-update

writeq-false-metric-update
writeq-true-metric-update

## writequeue effect on metric-add

writeq-false-metric-add
writeq-true-metric-add

File: benchmarks/simulate-lock-congestion/README.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
+runs mt-simulate-lock-congestion benchmarks,
+with parameters that mimic our problematic customer,
```
Comment (replay) [inline benchmarks/simulate-lock-congestion/README.md]:
```suggestion
with parameters that mimic problematic situations which we've seen in our production enrivonment,
```

File: cmd-dev/mt-simulate-memory-idx-lock-contention/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  queries straight from file queries-file.
+* issues AddOrUpdate calls from series-file.
+  typically updates, but upto new-series-percent is adds (except if we've already seen them), so the real add percentage is is <= new-series-percent.
+  partition is hashed from name
+  runs add-threads threads to add data to the index, each thread corresponds to 1 partition in the index.
+  the threads are fed data by the router, which routes data from the metrics generator, but limits it to rate of adds-per-sec // called adds but could these be updates?
+
+## obtaining the files
+
+TODO
```
Comment (replay) [inline cmd-dev/mt-simulate-memory-idx-lock-contention/README.md]:
```suggestion
Use the `mt-index-cat` utility to dump the metric names of the Metrictank index you want to test with into a file, then write a list of queries to issue into another file. Those two files can then be used as input by using the `series-file` and `queries-file` parameters.
```

File: cmd-dev/mt-simulate-memory-idx-lock-contention/main.go [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+		ticker := time.NewTicker(time.Second * 2)
+		for range ticker.C {
+			runtime.GC()
+		}
+	}()
+
+	testRun.Init(ctx)
+	go testRun.Run()
+
+	// TODO: this looks like a duplicate (see above): confirm we can remove this without affecting results
```
Comment (replay) [inline cmd-dev/mt-simulate-memory-idx-lock-contention/main.go]: I think this duplicate go func can be removed, I can't think of a reason why we'd need it twice so this was probably a mistake

**Pull request merged**